pallets/omnipool/src/types.rs

```diff
@@ -0,0 +1,628 @@
+		};
+		Some(Self { value, negative: sign })
+	}
+}
+
+/// Indicates whether hub asset changes the total issuance or not.
+#[derive(PartialOrd, PartialEq)]
+pub(super) enum HubAssetIssuanceUpdate {
+	AdjustSupply,
+	JustTransfer,
+}
+
+/// Indicates whether delta amount should be added or subtracted.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) enum BalanceUpdate<Balance> {
+	Increase(Balance),
+	Decrease(Balance),
+}
+
+impl<Balance: CheckedAdd + CheckedSub + PartialOrd + Copy + Default> BalanceUpdate<Balance> {
+	/// Merge two update together
+	pub(crate) fn merge(self, other: Self) -> Option<Self> {
+		self.checked_add(&other)
+	}
+}
+
+/// The addition operator + for BalanceUpdate.
+///
+/// Panics is overflows. Use checked_add for safe operation.
+///
+/// # Example
+///
+/// ```ignore
+/// assert_eq!(BalanceUpdate::Increase(100) + BalanceUpdate::Increase(100), BalanceUpdate::Increase(200));
+/// ```
+impl<Balance: CheckedAdd + CheckedSub + PartialOrd + Default> Add<Self> for BalanceUpdate<Balance> {
+	type Output = Self;
+
+	fn add(self, rhs: Self) -> Self::Output {
+		match (self, rhs) {
+			(Increase(a), Increase(b)) => BalanceUpdate::Increase(a + b),
+			(Decrease(a), Decrease(b)) => BalanceUpdate::Decrease(a + b),
+			(Increase(a), Decrease(b)) => {
+				if a >= b {
+					BalanceUpdate::Increase(a - b)
+				} else {
+					BalanceUpdate::Decrease(b - a)
+				}
+			}
+			(Decrease(a), Increase(b)) => {
+				if a >= b {
+					BalanceUpdate::Decrease(a - b)
+				} else {
+					BalanceUpdate::Increase(b - a)
+				}
+			}
+		}
+	}
+}
+
+/// Performs addition that returns `None` instead of wrapping around on overflow
+impl<Balance: CheckedAdd + CheckedSub + PartialOrd + Copy + Default> CheckedAdd for BalanceUpdate<Balance> {
+	fn checked_add(&self, v: &Self) -> Option<Self> {
+		match (self, v) {
+			(Increase(a), Increase(b)) => Some(BalanceUpdate::Increase(a.checked_add(b)?)),
+			(Decrease(a), Decrease(b)) => Some(BalanceUpdate::Decrease(a.checked_add(b)?)),
+			(Increase(a), Decrease(b)) => {
+				if a >= b {
+					Some(BalanceUpdate::Increase(a.checked_sub(b)?))
+				} else {
+					Some(BalanceUpdate::Decrease(b.checked_sub(a)?))
+				}
+			}
+			(Decrease(a), Increase(b)) => {
+				if a >= b {
+					Some(BalanceUpdate::Decrease(a.checked_sub(b)?))
+				} else {
+					Some(BalanceUpdate::Increase(b.checked_sub(a)?))
+				}
+			}
+		}
+	}
+}
+
+impl<Balance: Default> Default for BalanceUpdate<Balance> {
+	fn default() -> Self {
+		BalanceUpdate::Increase(Balance::default())
+	}
+}
+
+impl<Balance: Default> Deref for BalanceUpdate<Balance> {
+	type Target = Balance;
+
+	fn deref(&self) -> &Self::Target {
+		match self {
+			Increase(amount) | Decrease(amount) => amount,
+		}
+	}
+}
+
+#[macro_export]
+macro_rules! update_value {
+	( $x:expr, $y:expr) => {{
+		match &$y {
+			BalanceUpdate::Increase(amount) => $x.checked_add(&amount),
+			BalanceUpdate::Decrease(amount) => $x.checked_sub(&amount),
+		}
+	}};
```

Comment:
I notice that you solved a similar problem with `SimpleBalance` and `BalanceUpdate` (potentially negative number for unsigned ints) in different ways.
I would suggest building one abstraction. My personal opinion is that enum in combination with trait impl is the better combination (as opposed to using either of struct representation or this kind of macro for calculations).
Have you tried expressing this macro as an `Add` implementation?

Reply:
playground link for the add implementation:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c8da265a0ec8b4a085a576eeb3dbdc5e

Reply:
The issue i have with one abstraction is that `SimpleImbalance` is stored in storage.

and imbalance represents a number - it can be positive or negative. 

`BalanceUpdate` is used to represent that something is decreased/increased by that amount.

Using `BalanceUpdate` in storage  instead could be confusing to read imo, where you'd see `BalanceUpdate::Decrease(amount)`

Reply:
> playground link for the add implementation: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c8da265a0ec8b4a085a576eeb3dbdc5e

replaced the update_value macro with Add implementation as suggested.